File: central/app/services/certificates.py
from __future__ import annotations

import os
import subprocess
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


MAX_CERTIFICATE_FILES = 10
_CERTIFICATE_SUFFIX = ".crt"
_CERTIFICATE_UPDATE_TIMEOUT_SECONDS = 60


class CertificateManager:
    def __init__(
        self,
        storage_dir: Path,
        *,
        trust_target_dir: Path | None = None,
        update_command: str | None = None,
    ) -> None:
        self.storage_dir = storage_dir
        self.trust_target_dir = trust_target_dir or Path(
            os.getenv("RELAY_TRUST_TARGET_DIR", "/usr/local/share/ca-certificates/relay-centralizer")
        )
        self.update_command = update_command or os.getenv("RELAY_UPDATE_CA_CERTIFICATES", "update-ca-certificates")
        self.storage_dir.mkdir(parents=True, exist_ok=True)

    def snapshot(self) -> dict[str, Any]:
        return {
            "cert_dir": str(self.storage_dir),
            "max_files": MAX_CERTIFICATE_FILES,
            "files": self.list_files(),
        }

    def list_files(self) -> list[dict[str, Any]]:
        self.storage_dir.mkdir(parents=True, exist_ok=True)
        files = sorted((path for path in self.storage_dir.iterdir() if path.is_file()), key=lambda path: path.name.lower())
        return [self._file_payload(path) for path in files[:MAX_CERTIFICATE_FILES]]

    def save_uploaded_file(self, filename: str, content: bytes) -> dict[str, Any]:
        safe_name = self._sanitize_filename(filename)
        if Path(safe_name).suffix.lower() != _CERTIFICATE_SUFFIX:
            raise ValueError("only .crt certificate files are allowed")

        try:
            text = content.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ValueError("certificate must be UTF-8 PEM text") from exc

        text = self._normalize_certificate_text(text)
        if "-----BEGIN CERTIFICATE-----" not in text or "-----END CERTIFICATE-----" not in text:
            raise ValueError("certificate must be PEM encoded")

        existing = {path.name for path in self._all_files()}
        if safe_name not in existing and len(existing) >= MAX_CERTIFICATE_FILES:
            raise ValueError("only the first 10 certificate files are supported here")

        target = self.storage_dir / safe_name
        target.write_text(text, encoding="utf-8")
        self._install_trust_file(target)
        self._update_trust_store()
        return self._file_payload(target)

    def delete_file(self, filename: str) -> None:
        safe_name = self._sanitize_filename(filename)
        path = self.storage_dir / safe_name
        if not path.exists() or not path.is_file():
            raise FileNotFoundError(safe_name)
        path.unlink()
        (self.trust_target_dir / safe_name).unlink(missing_ok=True)
        self._update_trust_store()

    def _all_files(self) -> list[Path]:
        self.storage_dir.mkdir(parents=True, exist_ok=True)
        return sorted((path for path in self.storage_dir.iterdir() if path.is_file()), key=lambda path: path.name.lower())

    def _install_trust_file(self, source: Path) -> None:
        self.trust_target_dir.mkdir(parents=True, exist_ok=True)
        target = self.trust_target_dir / source.name
        target.write_text(source.read_text(encoding="utf-8"), encoding="utf-8")

    def _update_trust_store(self) -> None:
        result = subprocess.run(
            self.update_command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=_CERTIFICATE_UPDATE_TIMEOUT_SECONDS,
        )
        if result.returncode != 0:
            detail = (result.stderr or result.stdout or "").strip()
            raise RuntimeError(detail or f"{self.update_command} failed")

    def _normalize_certificate_text(self, text: str) -> str:
        return text.replace("\r\n", "\n").replace("\r", "\n")

    def _sanitize_filename(self, filename: str) -> str:
        safe_name = Path(str(filename or "").strip()).name
        if not safe_name or safe_name in {".", ".."}:
            raise ValueError("filename is required")
        return safe_name

    def _file_payload(self, path: Path) -> dict[str, Any]:
        return {
            "name": path.name,
            "size_bytes": path.stat().st_size,
            "modified_at": datetime.fromtimestamp(path.stat().st_mtime, tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        }

File: central/internal/services/certificates.go
package services

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const (
	MaxCertificateFiles = 10
	certSuffix          = ".crt"
	certUpdateTimeoutS  = 60
)

type CertManager struct {
	StorageDir     string
	TrustTargetDir string
	UpdateCommand  string
}

func NewCertManager(storageDir string) *CertManager {
	trustTarget := strings.TrimSpace(os.Getenv("RELAY_TRUST_TARGET_DIR"))
	if trustTarget == "" {
		trustTarget = "/usr/local/share/ca-certificates/relay-centralizer"
	}
	updateCmd := strings.TrimSpace(os.Getenv("RELAY_UPDATE_CA_CERTIFICATES"))
	if updateCmd == "" {
		updateCmd = "update-ca-certificates"
	}
	os.MkdirAll(storageDir, 0o755)
	return &CertManager{
		StorageDir:     storageDir,
		TrustTargetDir: trustTarget,
		UpdateCommand:  updateCmd,
	}
}

type CertFileInfo struct {
	Name       string `json:"name"`
	SizeBytes  int64  `json:"size_bytes"`
	ModifiedAt string `json:"modified_at"`
}

func (c *CertManager) Snapshot() map[string]interface{} {
	return map[string]interface{}{
		"cert_dir":  c.StorageDir,
		"max_files": MaxCertificateFiles,
		"files":     c.ListFiles(),
	}
}

func (c *CertManager) ListFiles() []CertFileInfo {
	os.MkdirAll(c.StorageDir, 0o755)
	entries, _ := os.ReadDir(c.StorageDir)
	var files []CertFileInfo
	for _, e := range entries {
		if e.IsDir() || len(files) >= MaxCertificateFiles {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		files = append(files, CertFileInfo{
			Name:       e.Name(),
			SizeBytes:  info.Size(),
			ModifiedAt: info.ModTime().UTC().Format(time.RFC3339),
		})
	}
	return files
}

func (c *CertManager) SaveUploadedFile(filename string, content []byte) (CertFileInfo, error) {
	safeName := sanitizeFilename(filename)
	if safeName == "" {
		return CertFileInfo{}, fmt.Errorf("filename is required")
	}
	if strings.ToLower(filepath.Ext(safeName)) != certSuffix {
		return CertFileInfo{}, fmt.Errorf("only .crt certificate files are allowed")
	}
	if !isUTF8(content) {
		return CertFileInfo{}, fmt.Errorf("certificate must be UTF-8 PEM text")
	}
	text := strings.ReplaceAll(strings.ReplaceAll(string(content), "\r\n", "\n"), "\r", "\n")
	if !strings.Contains(text, "-----BEGIN CERTIFICATE-----") || !strings.Contains(text, "-----END CERTIFICATE-----") {
		return CertFileInfo{}, fmt.Errorf("certificate must be PEM encoded")
	}

	existing := c.existingNames()
	if _, known := existing[safeName]; !known && len(existing) >= MaxCertificateFiles {
		return CertFileInfo{}, fmt.Errorf("only the first 10 certificate files are supported here")
	}

	target := filepath.Join(c.StorageDir, safeName)
	if err := os.WriteFile(target, []byte(text), 0o644); err != nil {
		return CertFileInfo{}, err
	}
	if err := c.installTrustFile(target); err != nil {
		return CertFileInfo{}, err
	}
	if err := c.updateTrustStore(); err != nil {
		return CertFileInfo{}, err
	}
	info, _ := os.Stat(target)
	return CertFileInfo{Name: safeName, SizeBytes: info.Size(), ModifiedAt: info.ModTime().UTC().Format(time.RFC3339)}, nil
}

func (c *CertManager) DeleteFile(filename string) error {
	safeName := sanitizeFilename(filename)
	path := filepath.Join(c.StorageDir, safeName)
	if err := os.Remove(path); err != nil {
		if os.IsNotExist(err) {
			return fmt.Errorf("%s: not found", safeName)
		}
		return err
	}
	os.Remove(filepath.Join(c.TrustTargetDir, safeName))
	return c.updateTrustStore()
}

func (c *CertManager) installTrustFile(src string) error {
	os.MkdirAll(c.TrustTargetDir, 0o755)
	content, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(c.TrustTargetDir, filepath.Base(src)), content, 0o644)
}

// ExecError represents a failure from running an OS-level command.
type ExecError struct {
	msg string
	Err error
}

func (e *ExecError) Error() string { return e.msg }

func (e *ExecError) Unwrap() error { return e.Err }

func (c *CertManager) updateTrustStore() error {
	cmd := exec.Command("sh", "-c", c.UpdateCommand)
	out, err := cmd.CombinedOutput()
	if err != nil {
		detail := strings.TrimSpace(string(out))
		if detail == "" {
			detail = fmt.Sprintf("%s failed", c.UpdateCommand)
		}
		return &ExecError{msg: detail, Err: err}
	}
	return nil
}

func (c *CertManager) existingNames() map[string]struct{} {
	entries, _ := os.ReadDir(c.StorageDir)
	m := make(map[string]struct{})
	for _, e := range entries {
		if !e.IsDir() {
			m[e.Name()] = struct{}{}
		}
	}
	return m
}
